SETUP/document_database.php

```diff
@@ -0,0 +1,273 @@
+        echo "  - generated table: \n";
+        echo "'$generated_table'\n";
+        echo "  - table from the documentation file: \n";
+        echo "'$table_in_file'\n";
+        exit(1);
+    }
+
+    // Check field definitions
+    $column_definitions_in_file = TableDocumentation::detect_columns_in_text($lines);
+    $column_definitions_from_table = $table_documentation->get_column_names();
+
+    $added_columns = array_diff($column_definitions_from_table, $column_definitions_in_file);
+    $deleted_columns = array_diff($column_definitions_in_file, $column_definitions_from_table);
+
+    if (count($added_columns) !== 0) {
+        foreach ($added_columns as $index => $column) {
+            echo "  - missing column definition for column '$column'\n";
+        }
+
+        exit(1);
+    }
+    if (count($deleted_columns) !== 0) {
+        foreach ($deleted_columns as $index => $column) {
+            echo "  - redundant column definition for column '$column' which does not exist in table\n";
+        }
+
+        exit(1);
+    }
+}
+
+/**
+ * Updates the markdown documentation file. Updating has two steps:
+ *
+ * 1. The documentation table is regenerated from the database information.
+ * 2. Missing column definitions are added, redundant column definitions for deleted columns are deleted.
+ *
+ * After update the file should pass verification.
+ *
+ * @param string $table_name the name of the table in the database, used for looking up columns
+ * @param string $file_path which file to verify
+ * @param string $display_name the title of the markdown file -- is not used
+ */
+function update_file_for_table(string $table_name, string $file_path, string $display_name) {
+    echo " - updating documentation for table '$table_name' in '$file_path'\n";
+
+    $columns = query_columns_for_table($table_name);
+
+    $table_documentation = new TableDocumentation($display_name, $columns);
+
+    $documentation_text = file_get_contents($file_path);
+    $lines = explode("\n", $documentation_text);
+
+    // Regenerate field table
+    $lines = TableDocumentation::replace_first_table($lines, [$table_documentation->generate_documentation_table()]);
+
+    // Add/remove field definitions
+    $column_definitions_from_table = $table_documentation->get_column_names();
+
+    $lines = TableDocumentation::update_column_definitions($lines, $column_definitions_from_table);
+
+    file_put_contents($file_path, implode("\n", $lines));
+}
+
+// ---------- Database functions ----------
+
+/**
+ * Runs a "DESCRIBE $table_name" query and returns the result.
+ */
+function query_columns_for_table(string $table_name): array {
+    $result = mysqli_query(DPDatabase::get_connection(), "DESCRIBE $table_name");
+
+    $rows = mysqli_fetch_all($result, MYSQLI_ASSOC);
+    mysqli_free_result($result);
+
+    return $rows;
+}
+
+/**
+ * Runs a "SHOW TABLES" query and returns the result.
+ */
+function query_table_names_in_current_database(): array {
```

Comment:
One problem with this approach is that not all tables are created in `db_schema.sql`, the best (and currently only?) example are the project id tables. Although I guess we will have to special case those in any case, so I agree with this solution. 

On a related note, are any project id tables created during CI? That would be useful for when we add the verification part to the CI.

Reply:
`projectID` tables should be the only things created in the code. The CI doesn't (yet?) create any of them but it should be easy to call `project_allow_pages()` somewhere so that one is created.

Reply:
To get the table name of a "live" project id table, we could: 

 * Create a table `projectIDxxxxxxxxxxxxx` in both CI and on the test server (with `project_allow_pages`) and then hardcode that table name in the function. 
 * Query the database for the most recently updated project id table name and use in the function.

Reply:
I think we're going to have to special case the `projectID` table description, regardless. Do we absolutely have to create a table to query, or use the most recently created `projectID` table? What I'm asking is, can we not just work with the information in `project_allow_pages`? If that can't be used, then the possible issue I see with querying the database for the most recently created `projectID` table is that on systems like the test server, there's the possibility that `project_allow_pages` might, in edge cases, have been updated since the most recent project creation.

Reply:
> What I'm asking is, can we not just work with the information in `project_allow_pages`?

Hmm, that is a good question, we _can_ parse the information from the `CREATE TABLE` query, however it is simpler just to query the database for that information. I think parsing the query would be more complex to implement and maintain. 

> If that can't be used, then the possible issue I see with querying the database for the most recently created `projectID` table is that on systems like the test server, there's the possibility that `project_allow_pages` might, in edge cases, have been updated since the most recent project creation.

Ah, then I think the best solution is to manually create and maintain (whenever necessary) a `projectIDxxxxxxxxxxxxx` table. It is more explicit and will be easier to debug.

Reply:
> Ah, then I think the best solution is to manually create and maintain (whenever necessary) a `projectIDxxxxxxxxxxxxx` table. It is more explicit and will be easier to debug.

"Create and maintain"? Or "Create and delete when the update is done"? I.e. do you really want a `projectID` table hanging around that really doesn't have anything to do with anything? It shouldn't have an entry in the `projects` table, and I would think that things would be cleaner if it were simply deleted. That way you don't have to worry about it creating problems anywhere else in the code.

Reply:
We can create it before running the command and delete it afterwards. By maintain I just meant recreating it if we ever change the create table query in code. :)

Reply:
I agree with this approach. We modify projectID table structures very very rarely.

Reply:
Resolved in c3b940f9214b4d9e0ce7909544b1700859bbfd4d.